### PR TITLE
Do roundtrip testing with external kinds in client-go TPR example

### PR DESCRIPTION
**Pull request by sttts**
This tests that our serialization machinery works for TPR types, i.e. without internal counterpart and without generated code.

/cc @nilebox

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/api/testing/roundtrip.go [hunk 1]
```diff
@@ -209,6 +214,32 @@ func roundTripToAllExternalVersions(t *testing.T, scheme *runtime.Scheme, codecF
...
+	}
+	typeAcc, err := meta.TypeAccessor(object)
+	if err != nil {
+		t.Fatalf("%q is not a TypeMeta and cannot be tested - add it to nonRoundTrippableInternalTypes: %v", externalGVK, err)
+	}
+
+	fuzzInternalObject(t, fuzzer, object)
+
+	externalGoType := reflect.TypeOf(object).PkgPath()
+	t.Logf("\tround tripping external type %v %v", externalGVK, externalGoType)
```
Comment (caesarxuchao) [inline staging/src/k8s.io/apimachinery/pkg/api/testing/roundtrip.go]: Is it necessary to log the externalGoType? We know where the unit tests are located.

File: staging/src/k8s.io/apimachinery/pkg/api/testing/roundtrip.go [hunk 2]
```diff
@@ -209,6 +214,32 @@ func roundTripToAllExternalVersions(t *testing.T, scheme *runtime.Scheme, codecF
...
+
+	typeAcc.SetKind(externalGVK.Kind)
+	typeAcc.SetAPIVersion(externalGVK.GroupVersion().String())
+
+	roundTrip(t, scheme, json.NewSerializer(json.DefaultMetaFactory, scheme, scheme, false), object)
```
Comment (caesarxuchao) [inline staging/src/k8s.io/apimachinery/pkg/api/testing/roundtrip.go]: This is a round-trip of marshalling/unmarshalling. Could you update the comment on `roundTrip`? I think it used to mean internal->external->internal.

File: staging/src/k8s.io/client-go/examples/third-party-resources/apis/tpr/v1/types_test.go [hunk 3]
```diff
@@ -17,12 +17,45 @@ limitations under the License.
...
+			c.FuzzNoCustom(obj)
+			obj.Items = make([]Example, c.Intn(10))
+			for i := range obj.Items {
+				c.Fuzz(&obj.Items[i])
+			}
+		},
+	}
+}
+
+func TestRoundTrip(t *testing.T) {
```
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/examples/third-party-resources/apis/tpr/v1/types_test.go]: I think the purpose is testing the way we build the scheme/codecs are effective, i.e., the following three lines, right? Can we explicitly say that in the comment?

**Review by caesarxuchao — COMMENTED** [review]
A few nits. Otherwise lgtm.

File: staging/src/k8s.io/apimachinery/pkg/api/testing/roundtrip.go (hunk 1 above)
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/api/testing/roundtrip.go]: It's just copying what the func above does for internal types. Would like to keep it symmetric.

File: staging/src/k8s.io/client-go/examples/third-party-resources/apis/tpr/v1/types_test.go (hunk 3 above)
Comment (sttts) [inline staging/src/k8s.io/client-go/examples/third-party-resources/apis/tpr/v1/types_test.go]: done

File: staging/src/k8s.io/apimachinery/pkg/api/testing/roundtrip.go (hunk 2 above)
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/api/testing/roundtrip.go]: done

**Comment by sttts** [comment]
Addressed the comments. Left in the `t.Logf` to stay symmetric with internal types.

**Comment by sttts** [comment]
Self-approved by https://github.com/kubernetes/kubernetes/pull/45543#pullrequestreview-37087513 as @caesarxuchao is owner.

**Pull request merged**